### ISSUE TITLE
Comment in "record stock" is ignored

### ISSUE THREAD

**Issue by WebFreak001** [openboxes/openboxes#1383]
When in a product clicking the "Record stock" button and editing the comment in the inline table, it will never get added to the item when saving.

Manually clicking the arrow in the "actions" column in the stock card and editing item and adding a comment in the big popup window works though.

**Comment by jmiranda** [comment]
The comment fields on the Record Stock page are saved at the transaction / transaction entry level so they’re currently only displayed on the Stock History tab. 

Where are you expecting to see these comments? Can you provide an example use case?

**Comment by WebFreak001** [comment]
this is the current stock cards with the comments column:
![current stock cards](https://wfr.moe/f6ItxH.png)

this is the record stock UI:
![record stock ui](https://wfr.moe/f6ItK9.png)

now the columns almost exactly match up, so I would expect editing the product there to take over the comment of the product in the stock card. Especially because this would be much more useful than a transaction comment imo. However putting in comments inside these fields don't assign anything, so I guess they are put in as the transaction comment.

I think for a transaction comment it would be more convenient and clearer if there was one single input field to describe the whole transaction with all added items and have these comments input fields instead edit the product comments.

**Comment by jmiranda** [comment]
> now the columns almost exactly match up, so I would expect editing the product there to take over the comment of the product in the stock card.

Ah yeah, I can see where that might be confusing. The comment on the In Stock tab is associated with a comment field on the Inventory Item. You can edit that comment using the Edit Item action.
![Amoxicillin  500mg tablet](https://user-images.githubusercontent.com/1162935/71654704-c95c4380-2cf8-11ea-86be-5a1d7580887a.png).

Does that make things better or worse? The problem with displaying the transaction or transaction entry comment in the In Stock tab is the fact that you generally have multiple transactions associated with a single line. So which comment would you use? Most recent? 

Can you give me some examples of what you want to put in the comment field?

**Comment by WebFreak001** [comment]
My use case: I don't have that many items so instead of using generic products to group a lot of products, I create one product -- for example called `USB Type A <-> Micro Type B cable` -- and have multiple lines in the stock card. If I bothered to write it down, I put in things like `red, 1.5m`.

Basically right now if I have a product where all items are functionally nearly identical and can be interchanged, I store all items in that one product and use the comments tab to write down details. Also when I don't put any ID or sticker on it and check the bin where it is stored, I know which one I actually took out of the system using the comment.

I only personally use openboxes at home to manage my electronics stuff but I think that's actually a really great use case of openboxes too.

**Comment by jmiranda** [comment]
Awesome. You're basically looking for product variants, which we don't support just yet. And yeah you were on the right track about product groups. That's essentially what that feature is for but it's a bit of a pain to work with so I like your approach for now. 

Will editing the inventory item comment work for your needs?

> Basically right now if I have a product where all items are functionally nearly identical and can be interchanged, I store all items in that one product and use the comments tab to write down details.

In the future, we could set the comment on the inventory item when you include a comment in a line on the Record Stock or Import Inventory feature. 

Would that be useful? 

Since both features can be used for existing products, we'd need to figure out how to handle comment overrides. The expected behavior could be "overwrite inventory item comment if one exists, ignore if it is blank". That way you don't overwrite the comment when the Record Stock or Import Inventory comment is blank.

**Comment by WebFreak001** [comment]
I would suggest to simply:
- copy comments from table you see in stock view into record stock inputs
- on submit, for each entry line:
	- if comment isn't changed, don't change the comment in the item entry
	- if comment is changed, change the item entry along with everything else

I would have only one single comments field (the comments field at the top which already exists I guess?) which stores comments about the transaction you are creating. This is much more like what the UI feels like it has been designed for.

Otherwise how it is right now I don't really get why there is a big comments field at the top and separate comment fields in each line. My understanding of a transaction is that it is the whole changeset of all items I'm changing, and thus I expect the single comments block at the top of the page to be the transaction comment.

![record stock header](https://wfr.moe/f6IbO4.png)
I would expect this big text area to be the transaction comment instead ^

**Comment by jmiranda** [comment]
> copy comments from table you see in stock view into record stock inputs
Oh so if you have "red, 1.5 m" as a comment on the inventory item you want that to be placed in the comments field of the Record Stock line item?

> I would have only one single comments field (the comments field at the top which already exists I guess?) which stores comments about the transaction you are creating. This is much more like what the UI feels like it has been designed for. Otherwise how it is right now I don't really get why there is a big comments field at the top and separate comment fields in each line. 

There are two ways to comment on any transaction. One is a general comment about the transaction (the "big text area" on the Record Stock page). Then you might also want to comment on each line ("found 2 boxes in bulk storage" or "damaged 1 ea - person b ran it over with the forklift"). 

> My understanding of a transaction is that it is the whole changeset of all items I'm changing, and thus I expect the single comments block at the top of the page to be the transaction comment.

That is correct. Both the transaction and transaction entry comments are displayed in the stock history.

**Comment by jmiranda** [comment]
I don't remember the conclusion of this ticket, but I have two expectations:
* the comments from the Record Stock transaction should show up in Stock History (at the very least). If we want to include the comments from the latest transaction (or all active transactions) on the Stock Card, that would be ok but would be a new feature request.
* the inventory item (product lot) comments should be displayed on the Stock Card page next to the lot number (i.e. using the post-it note icon and onHover or title attribute)

I could have sworn that was already implemented (or the comments field was displayed as a column on the Stock Card). 

If it's not there, do we know when it was removed?

**Comment by ewaterman** [comment]
> the comments from the Record Stock transaction should show up in Stock History

Yes this is what happens currently. I don't think we need to show record stock comments in the stock card as it wouldn't be clear where the comments are coming from. Stock History seems like a clear enough place to put that.

> the inventory item (product lot) comments should be displayed on the Stock Card page next to the lot number

Lot comment is not visible currently on stock card. You're right that it did exist but it was removed.

Here is it being added (in 2019): https://github.com/openboxes/openboxes/pull/1122/files#diff-75df08df9cdba25a3b7d3349c3270c7064a0b521271e953a9c1d024437262843R22

And here is it being removed (in 2021): https://github.com/openboxes/openboxes/pull/2766/files#diff-75df08df9cdba25a3b7d3349c3270c7064a0b521271e953a9c1d024437262843L23

This is the ticket where the comments column was requested to be removed: https://pihemr.atlassian.net/browse/OBPIH-4073

Unfortunately there's no discussion about it, just the request to remove the column.

It's probably fine to continue not showing the comments column on the stock card, but I'd say it's worth showing comments on the "manage lot numbers" page since you're directly editing lots there and can still provide a comment.


It should be really simple to add to `/views/inventoryItem/_showLotNumbers.gsp`:
- Add a column "Comments" (can use `default.comments.label` for localization)
- For every inventory item, add the value of `inventoryItem?.expirationDate` to the column for that row
